uniffi_bindgen/src/bindings/kotlin/templates/Helpers.kt

```diff
@@ -28,3 +28,73 @@ inline fun <T : FFIObject, R> T.use(block: (T) -> R) =
             // swallow
         }
     }
+
+typealias Handle = Long
+class ConcurrentHandleMap<T>(
+    private val leftMap: MutableMap<Handle, T> = mutableMapOf(),
+    private val rightMap: MutableMap<T, Handle> = mutableMapOf()
+) {
+    private val lock = java.util.concurrent.locks.ReentrantLock()
+    private val currentHandle = AtomicLong(0L)
+    private val stride = 1L
+
+    fun insert(obj: T): Handle =
+        lock.withLock {
+            rightMap[obj] ?:
+                currentHandle.getAndAccumulate(stride) { a, b -> a + b }
+                    .also { handle ->
+                        leftMap[handle] = obj
+                        rightMap[obj] = handle
+                    }
+            }
+
+    fun <R> callWithResult(handle: Handle, fn: (T) -> R): R =
+        lock.withLock {
+            val obj = leftMap[handle] ?: throw RuntimeException("Panic: handle not in handlemap")
+            fn.invoke(obj)
+        }
```

Comment:
Right, good call. And things should be safe here once we've got a reference to the object out of the handlemap, because  the GC will keep it alive during the call.

Reply:
The `ConcurrentHandleMap` in Rust sounds like it has this same implementation and similar deadlock weakness.